Add support library fragment watcher

AndreasBoehm opened the pull request:
Because there are some projects that can't migrate to AndroidX yet and my main project is one of those I figured that it might make sense to share this code with the community to allow everyone to use leakcanary 2 with the good old support library.

To watch leaking support library fragments and their views the following needs to be added to the **build.gradle**.
```
  debugImplementation 'com.squareup.leakcanary:leakcanary-android:2.0-beta-4'
  debugImplementation 'com.squareup.leakcanary:leakcanary-object-watcher-android-support-fragments:2.0-beta-4'
```

To test this PR you need to add the support-fragments project to the **build.gradle**.
```
 debugImplementation project(':leakcanary-object-watcher-android-support-fragments')
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/square/leakcanary?pullRequest=1611) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/square/leakcanary?pullRequest=1611) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/square/leakcanary?pullRequest=1611) it.</sub>

Armaxis commented:
This looks useful to me; @pyricau what do you think?

Quick (somewhat related) question: @AndreasBoehm do you see AndroidX dependencies being pulled into project when using LeakCanary? I'm wondering if developers who still use Support Lib end up pulling the AndroidX classes via transitive dependency through LeakCanary.

pyricau commented on leakcanary-object-watcher-android-support-fragments/src/main/java/leakcanary/internal/AndroidSupportFragmentDestroyWatcher.kt:
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
```
2019 ;) ?

AndreasBoehm replied:
Copy & paste 🤦‍♂️

AndreasBoehm replied:
Updated

pyricau commented:
Can you also add this module here: https://github.com/square/leakcanary/blob/master/leakcanary-android-core/build.gradle#L7

Essentially this will remove the need to add `  debugImplementation 'com.squareup.leakcanary:leakcanary-object-watcher-android-support-fragments:2.0-beta-4'
`. The code for both Android X and support lib will be included  but not loaded, and then at runtime we load it if and only if the right classes are available.

Side note: we should make sure this doesn't cause issues with jetifier

AndreasBoehm commented on leakcanary-object-watcher-android/src/main/java/leakcanary/internal/FragmentDestroyWatcher.kt:
```diff
@@ -57,6 +61,17 @@ internal object FragmentDestroyWatcher {
+          watcherConstructor.newInstance(objectWatcher, configProvider) as (Activity) -> Unit
+      )
+    }
+
```
I actually wanted to do this but forget about it after everything worked as expected. Sorry!

AndreasBoehm replied:
Addressed within my latest commit

AndreasBoehm commented:
Thanks for accepting my PR! I would love to adjust it right now but i am on holidays and i don’t have my computer with me.

I can adjust it in roughly two weeks if this is fine for you?
Fine for me if you want to adjust it on your own.

I wasn’t sure about adding the support library watcher to the core package as i didn’t want to add code for a legacy library to the project. But i agree that it would make things easier for users of the library.

AndreasBoehm commented:
> This looks useful to me; @pyricau what do you think?
> 
> Quick (somewhat related) question: @AndreasBoehm do you see AndroidX dependencies being pulled into project when using LeakCanary? I'm wondering if developers who still use Support Lib end up pulling the AndroidX classes via transitive dependency through LeakCanary.

Thats a good question!
I did not check if this is pulling in the AndroidX deps.

pyricau commented:
I'm ok with waiting, would love to give you the opportunity to finish this.

Once this is ready we should probably use it  on a jetified project and make sure Gradle won't complain.

AndreasBoehm commented:
Thank you!

Will work on it after my holidays.

Talk to you soon.

AndreasBoehm commented on build.gradle:
```diff
@@ -34,6 +34,7 @@ buildscript {
       mockito_kotlin: 'com.nhaarman:mockito-kotlin-kt1.1:1.5.0',
       okio: 'com.squareup.okio:okio:2.2.2',
       robolectric   : 'org.robolectric:robolectric:4.0-alpha-3',
+      android_support: 'com.android.support:support-v4:28.0.0'
```
moved :)

AndreasBoehm commented:
Hi @pyricau,

I addressed your comments but I was not able to test the library on a jetified project.
For my previous tries, I added the library as a source dependency to my project but to test with jetifier I would need to add it via maven.
Could you help me to install the library to my local maven repository?

@Armaxis I will check later today if leakcanary pulled in AndroidX dependencies into my project.

pyricau commented:
Try this (changing the path to your local repo):

```
./gradlew uploadArchives -PSNAPSHOT_REPOSITORY_URL=file:///Users/py/.m2/repository
```

pyricau commented:
Updated the docs to document this: https://square.github.io/leakcanary/dev-env/#deploying-locally

pyricau commented:
I checked out the branch, deployed locally, added the support library only to a project then support library + jetifier and it's all been working, no issue.

I'm going to merge :)
